File: 092.Reverse Linked List II.java
/*
    Reverse a linked list from position m to n. Do it in-place and in one-pass.

    For example:
    Given 1->2->3->4->5->NULL, m = 2 and n = 4,

    return 1->4->3->2->5->NULL.

    Note:
    Given m, n satisfy the following condition:
    1 ≤ m ≤ n ≤ length of list.
    */

    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode(int x) { val = x; }
     * }
     */

    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode firstPoint = dummy, endPoint = dummy;
        if (head == null || m == n)
            return head;
        for (int i=0; i<=n; i++) {
            if (i == m-1)
                firstPoint = endPoint;
            endPoint = endPoint.next;
        }
        ListNode one = firstPoint.next;
        ListNode two = one.next;
        ListNode three;
        ListNode end = firstPoint.next;
        while (two != null && two != endPoint) {
            three = two.next;
            two.next = one;
            one = two;
            two = three;
        }
        firstPoint.next.next = endPoint;
        firstPoint.next = one;

        return dummy.next;
    }

File: LeetCode/092.Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        cur = dummy
        i = 0
        while i < m - 1:
            cur = cur.next
            i += 1
        start = last = cur
        end = cur = cur.next
        while i < n:
            tmp = cur.next
            cur.next = last
            last = cur
            cur = tmp
            i += 1
        start.next = last
        end.next = cur
        return dummy.next
